modules/core/keeper/msg_server.go

```diff
@@ -751,7 +751,54 @@ func (k Keeper) ChannelUpgradeInit(goCtx context.Context, msg *channeltypes.MsgC
 
 // ChannelUpgradeTry defines a rpc handler method for MsgChannelUpgradeTry.
 func (k Keeper) ChannelUpgradeTry(goCtx context.Context, msg *channeltypes.MsgChannelUpgradeTry) (*channeltypes.MsgChannelUpgradeTryResponse, error) {
-	return nil, nil
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	module, _, err := k.ChannelKeeper.LookupModuleByChannel(ctx, msg.PortId, msg.ChannelId)
+	if err != nil {
+		ctx.Logger().Error("channel upgrade try failed", "port-id", msg.PortId, "error", errorsmod.Wrap(err, "could not retrieve module from port-id"))
+		return nil, errorsmod.Wrap(err, "could not retrieve module from port-id")
+	}
+
+	cbs, ok := k.Router.GetRoute(module)
+	if !ok {
+		ctx.Logger().Error("channel upgrade try failed", "port-id", msg.PortId, "error", errorsmod.Wrapf(porttypes.ErrInvalidRoute, "route not found to module: %s", module))
+		return nil, errorsmod.Wrapf(porttypes.ErrInvalidRoute, "route not found to module: %s", module)
+	}
+
+	upgrade, err := k.ChannelKeeper.ChanUpgradeTry(ctx, msg.PortId, msg.ChannelId, msg.ProposedUpgradeConnectionHops, msg.UpgradeTimeout, msg.CounterpartyProposedUpgrade, msg.CounterpartyUpgradeSequence, msg.ProofChannel, msg.ProofUpgrade, msg.ProofHeight)
```

Comment:
Note: if this contained state changes, we would need to be doing a cache context around it. I wonder if we should do this defensively?

Reply:
Food for thought

It looks like we have a pattern:
```go
writeFn() := cacheCtx()
err := funcCall()
if isUpgradError() {
    return Abort()
}

writeFn()
```

I wonder if we can abstract this into a helper function? Like `handleUpgrade()`? 

I guess the OnChanUpgradeTry func would need to provide:
```go
func() {
    err := OnChanUpgradeTry()
    if err != nil {
        return types.NewUpgradeError(err)
    }
    
    return nil
}
```

Reply:
I was thinking something similar actually,I think it would be nice if we could do something like

```golang
defer handleAbort(...)
err := OnChanUpgradeTry(...)
...
```

I think we can leave this for now though and maybe consider it later.

Reply:
Agree using `cacheCtx` here provides more defence in depth but restoring the channel in `AbortUpgrade` is supposed to delete the upgrade and set the channel back to its open and operational state.

Let's open an issue to discuss this later. @colin-axner made great points about INIT events being emitted as false positives in an error situation from [here](https://github.com/cosmos/ibc-go/blob/04-channel-upgrades/modules/core/04-channel/keeper/upgrade.go#L134)

Reply:
#3823

Reply:
#3824